### DISCUSSION TITLE
Issue with mariadb and galera cluster

### DISCUSSION THREAD

**Discussion by jydb2** [keycloak/keycloak#17026] [Archive]
Hello.
We are using keycloak version 20.0.3 using docker (quarkus).
My database infrastructure is based on 3 nodes running mariadb 10.5 and galera 4 (26.4.11).
My database cluster is up and running. 

At first launched we had an issue about mandatory primary key.

>  INFO  [org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusJpaUpdaterProvider] (main) Initializing database schema. Using changelog META-INF/jpa-changelog-master.xml
>  WARN  [org.mariadb.jdbc.message.server.ErrorPacket] (main) Error: 1173-42000: This table type requires a primary key
>  INFO  [org.jgroups.protocols.pbcast.GMS] (keycloak-cache-init) 5e4cb6b13b8c-42049: no members discovered after 2004 ms: creating cluster as coordinator
>  INFO  [org.infinispan.CLUSTER] (keycloak-cache-init) ISPN000094: Received new cluster view for channel ISPN: [5e4cb6b13b8c-42049|0] (1) [5e4cb6b13b8c-42049]
>  INFO  [org.infinispan.CLUSTER] (keycloak-cache-init) ISPN000079: Channel `ISPN` local address is `5e4cb6b13b8c-42049`, physical addresses are `[172.18.0.2:55457]`
>  INFO  [org.infinispan.CLUSTER] (main) ISPN000080: Disconnecting JGroups channel `ISPN`
>  ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (production) mode
>  ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to update database
>  ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: (conn=24419) This table type requires a primary key [Failed SQL: (1173) CREATE TABLE gv_keycloak.DATABASECHANGELOG (ID VARCHAR(255) NOT NULL, AUTHOR VARCHAR(255) NOT NULL, FILENAME VARCHAR(255) NOT NULL, DATEEXECUTED datetime NOT NULL, ORDEREXECUTED INT NOT NULL, EXECTYPE VARCHAR(10) NOT NULL, MD5SUM VARCHAR(35) NULL, `DESCRIPTION` VARCHAR(255) NULL, COMMENTS VARCHAR(255) NULL, TAG VARCHAR(255) NULL, LIQUIBASE VARCHAR(20) NULL, CONTEXTS VARCHAR(255) NULL, LABELS VARCHAR(255) NULL, DEPLOYMENT_ID VARCHAR(10) NULL)]
>  ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: (conn=24419) This table type requires a primary key

If we are adding missing mandatory keys, keycloak is running well. 
But from my point of view changing data model is not the right thing to do.
A workaround on my mariadb nodes was to set innodb_force_primary_key=0 I don't know if we'll have an impact in a near future.

Why not using primary keys on tables (seems to be mandatory with galera) ?. 
Could we create primary keys on tables and not having issue in the future (upgrade to version 21 for example) 
Is galera cluster really supported and when it will be no longer ?  
I found this about new store : https://github.com/keycloak/keycloak/discussions/10404

Regards,

Jean-Yves?

**Comment by vramik** [comment]
Hello @jydb2, it seems that it might be similar issue as described here: https://issues.redhat.com/browse/KEYCLOAK-4928 

The more details why databasechangelog table doesn't have a primary key might be found here https://liquibase.jira.com/browse/CORE-1909

**Reply by jydb2** [reply]
Thank you for your reply. 
Is galera still supported for a long time or should we consider moving to another database ?.
Regards.

**Reply by vramik** [reply]
With new store there will be first and second class databases. For more information, please see: https://www.keycloak.org/2022/02/dbs.html

**Comment by darius-m** [comment]
Using `wsrep_certify_nonPK=1` as a setting may be an alternative to not requiring having any primary keys set, as it only applies to Galera transactions. Not sure if one of the options is worse by default.

Regarding which databases should be used, I believe only postgres has prime tier support, while mysql-derived variants are more best-effort.

**Reply by jydb2** [reply]
Thanks for your reply.
We have to find the best way to implement keycloak from now.
